src/daft-core/src/array/ops/utf8.rs

```diff
@@ -63,15 +63,21 @@ where
 
 impl Utf8Array {
     pub fn endswith(&self, pattern: &Utf8Array) -> DaftResult<BooleanArray> {
-        self.binary_broadcasted_compare(pattern, |data: &str, pat: &str| data.ends_with(pat))
+        self.binary_broadcasted_compare(pattern, |data: &str, pat: &str| Ok(data.ends_with(pat)))
     }
 
     pub fn startswith(&self, pattern: &Utf8Array) -> DaftResult<BooleanArray> {
-        self.binary_broadcasted_compare(pattern, |data: &str, pat: &str| data.starts_with(pat))
+        self.binary_broadcasted_compare(pattern, |data: &str, pat: &str| Ok(data.starts_with(pat)))
     }
 
     pub fn contains(&self, pattern: &Utf8Array) -> DaftResult<BooleanArray> {
-        self.binary_broadcasted_compare(pattern, |data: &str, pat: &str| data.contains(pat))
+        self.binary_broadcasted_compare(pattern, |data: &str, pat: &str| Ok(data.contains(pat)))
+    }
+
+    pub fn match_(&self, pattern: &Utf8Array) -> DaftResult<BooleanArray> {
+        self.binary_broadcasted_compare(pattern, |data: &str, pat: &str| {
+            Ok(regex::Regex::new(pat)?.is_match(data))
```

Comment:
https://docs.rs/regex/latest/regex/struct.Regex.html#method.new
> Note that regex compilation tends to be a somewhat expensive process, and unlike higher level environments, compilation is not automatically cached for you. One should endeavor to compile a regex once and then reuse it. For example, it’s a bad idea to compile the same regex repeatedly in a loop.

You should try moving `regex::Regex::new(pat)` outside of the callback if possible

Reply:
great catch!